lib/ecto/pools/ownership.ex

```diff
@@ -0,0 +1,202 @@
+      Process.flag(:trap_exit, true)
+      {:ok, %{module: module,
+              pool: pool,
+              ets: ets,
+              owner: nil,
+              ref: nil,
+              checkout: nil,
+              strategy: nil}}
+    end
+
+    def handle_call({:ownership_checkout, strategy, timeout}, {pid, _ref}, s) do
+      case s.module.checkout(s.pool, timeout) do
+        {:ok, worker, {module, conn}, :default, _queue_time} ->
+          mode = if strategy,
+                   do: strategy.ownership_checkout(module, conn),
+                 else: :default
+
+          checkout = {worker, {module, conn}, mode}
+
+          s = monitor_owner(pid, s)
+          s = %{s | strategy: strategy, checkout: checkout}
+          {:reply, {:ok, checkout}, s}
+        {:error, _} = error ->
+          {:stop, :normal, error, s}
+      end
+    end
+
+    def handle_call({:ownership_checkin, timeout}, _from, s) do
+      checkin(timeout, s)
+      {:stop, :normal, :ok, s}
+    end
+
+    def handle_info({:DOWN, ref, :process, pid, _reason}, %{ref: ref, owner: pid} = s) do
+      checkin(@timeout, s)
+      {:stop, :normal, s}
+    end
+
+    def terminate(_reason, s) do
+      if s.owner do
+        :ets.delete(s.ets, s.owner)
+      end
+    end
+
+    defp checkin(timeout, s) do
+      {worker, {module, conn}, _} = s.checkout
+      if s.strategy, do: s.strategy.ownership_checkin(module, conn)
+      s.module.checkin(s.pool, worker, timeout)
+    end
+
+    defp monitor_owner(owner, %{owner: nil, ref: nil} = s) do
+      ref = Process.monitor(owner)
+      %{s | owner: owner, ref: ref}
+    end
+  end
+
+  defmodule ServerSup do
+    use Supervisor
+
+    def start_link(name, pool, pool_name, ets) do
+      Supervisor.start_link(__MODULE__, {pool, pool_name, ets}, name: name)
+    end
+
+    def init({pool, pool_name, ets}) do
+      children = [
+        worker(Server, [pool, pool_name, ets], restart: :temporary)
+      ]
+      supervise(children, strategy: :simple_one_for_one)
+    end
+
+    def new_owner(name) do
+      Supervisor.start_child(name, [])
+    end
+  end
+
+  defmodule Sup do
+    use Supervisor
+
+    def start_link(connection, opts) do
+      Supervisor.start_link(__MODULE__, {connection, opts})
+    end
+
+    def init({connection, opts}) do
+      {pool, opts} = Keyword.pop(opts, :ownership_pool)
+      name         = Keyword.fetch!(opts, :pool_name)
+      pool_name    = Module.concat(name, Inner)
+      pool_opts    = Keyword.put(opts, :pool_name, pool_name)
+      sup_name     = Module.concat(name, Elixir.ServerSup)
+
+      :ets.new(name, [:named_table, :public, read_concurrency: true])
+      :ets.insert(name, {:metadata, sup_name, pool_name, pool})
+
+      children = [
+        supervisor(ServerSup, [sup_name, pool, pool_name, name]),
+        supervisor(pool, [connection, pool_opts])
+      ]
+      supervise(children, strategy: :rest_for_one)
+    end
+  end
+
+  @behaviour Ecto.Pool
+  @timeout 5_000
```

Comment:
When we call `__pool__`, the third argument is a timeout. Let's use that instead of the hardcoded value. In fact, there is no need to expect a timeout value in the `ownership_checkout` and checkin below.